pkg/security/admission/admission.go

```diff
@@ -0,0 +1,318 @@
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return kclient.SecurityContextConstraints().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kclient.SecurityContextConstraints().Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&kapi.SecurityContextConstraints{},
+		store,
+		0,
+	)
+	reflector.Run()
+
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  kclient,
+		store:   store,
+	}
+}
+
+// Admit determines if the pod should be admitted based on the requested security context
+// and the available SCCs.
+//
+// 1.  Find all the SCCs that the service account or user has access to
+// 2.  Full resolve each SCC
+//     1.  Determine if the user/selinux strategies requires pre-allocated values
+//     2.  Retrieve pre-allocated values from the namespace and set them on the strategies
+// 3.  Create the set of providers from the fully configured SCCs
+// 4.  Generate and validate each container's SC against the set of SCC providers.  If any container
+//     fails to validate then the entire request is rejected
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// get all constraints that are usable by the user
+	matchedConstraints, err := getMatchingSecurityContextConstraints(c.store, a.GetUserInfo())
+	if err != nil {
+		return kadmission.NewForbidden(a, err)
+	}
+	// if there is a service account get the constraints for the service account
+	if len(pod.Spec.ServiceAccount) > 0 {
+		serviceAccount, err := getServiceAccount(c.client, a.GetNamespace(), pod)
+		if err != nil {
+			return kadmission.NewForbidden(a, err)
+		}
+
+		userInfo := serviceaccount.UserInfo(a.GetNamespace(), serviceAccount.Name, string(serviceAccount.UID))
+		saConstraints, err := getMatchingSecurityContextConstraints(c.store, userInfo)
+		if err != nil {
+			return kadmission.NewForbidden(a, err)
+		}
+		matchedConstraints = append(matchedConstraints, saConstraints...)
+	}
+
+	providers := make(map[string]scc.SecurityContextConstraintsProvider, 0)
+	// namespace is declared here for reuse but we will not fetch it unless required by
+	// the matched constraints
+	var namespace *kapi.Namespace
+
+	// set pre-allocated values on constraints.  By placing them in a map we also de-dupe the constraints
+	// that match the sa and user.
+	for _, constraint := range matchedConstraints {
+		if requiresPreAllocatedUIDRange(constraint) {
+			if namespace == nil {
+				namespace, err = c.client.Namespaces().Get(a.GetNamespace())
+				if err != nil {
+					return kadmission.NewForbidden(a, err)
```

Comment:
not sure you should return here... this means that this constraint had an error, but another constraint might not.